src/database.c

```diff
@@ -493,3 +555,159 @@ _add_to_db(ProfMessage* message, char* type, const Jid* const from_jid, const Ji
+            "SELECT `from_jid`, `to_jid`, `from_resource`, `to_resource`, `message`, `timestamp`, `type`, `stanza_id`, `archive_id`, `replace_id` AS `replaces_stanza_id`, `encryption`, `marked_read`"
+            "FROM `ChatLogs`;";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to copy V2 table.");
+        goto cleanup;
+    }
+
+    // Save original message and DB ID in replacement messages
+    // Note: sender check is required due to #1898 (24d0030) since previous messages might be affected.
+    query = "UPDATE `ChatLogsV2` AS A "
+            "SET `original_message` = B.`message`, "
+            "`replaces_db_id` = B.`id` "
+            "FROM `ChatLogs` AS B "
+            "WHERE A.`replaces_stanza_id` IS NOT NULL AND A.`replaces_stanza_id` != '' AND A.`replaces_stanza_id` = B.`stanza_id` "
+            "AND A.`from_jid` = B.`from_jid` AND A.`to_jid` = B.`to_jid`;";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to add original messages to replacement messages in V2 table.");
+        goto cleanup;
+    }
+
+    // Set original message to redacted version
+    // Note: sender check is required due to #1898 (24d0030) since previous messages might be affected.
+    query = "UPDATE `ChatLogsV2` AS A "
+            "SET `message` = COALESCE(B.`message`, A.`message`)"
+            "FROM `ChatLogsV2` AS B "
+            "WHERE B.`replaces_db_id` IS NOT NULL AND B.`replaces_db_id` != 0 AND A.`id` = B.`replaces_db_id` "
+            "AND A.`from_jid` = B.`from_jid` AND A.`to_jid` = B.`to_jid`;";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to change messages in V2 table.");
+        goto cleanup;
+    }
+
+    // Cleanup LMC as we only keep "original message" in them so we don't store the same data twice
+    query = "UPDATE `ChatLogsV2` "
+            "SET `message` = NULL "
+            "WHERE `replaces_stanza_id` != '';";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to change messages in V2 table.");
+        goto cleanup;
+    }
+
+    // Rename original table to new name
+    query = "ALTER TABLE ChatLogs RENAME TO ChatLogs_bck;";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to rename old table.");
+        goto cleanup;
+    }
+
+    // Set V2 table as main
+    query = "ALTER TABLE ChatLogsV2 RENAME TO ChatLogs;";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to rename new table.");
+        goto cleanup;
+    }
+
+    // TODO: drop old table (_bck). Only after sufficient testing this part should be added
```

Comment:
Before merging this should be activated. We dont want the db file to become too big.